File: fop-core/src/main/java/org/apache/fop/pdf/PDFReference.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.pdf;

import java.io.OutputStream;
import java.lang.ref.Reference;
import java.lang.ref.SoftReference;

/**
 * Class representing a PDF object reference. The object holds a soft reference to the actual
 * PDF object so the garbage collector can free the object if it's not referenced elsewhere. The
 * important thing about the class is the reference information to the actual PDF object in the
 * PDF file.
 */
public class PDFReference implements PDFWritable {

    private PDFObjectNumber objectNumber;
    private int generation;

    private Reference<PDFObject> objReference;

    /**
     * Creates a new PDF reference.
     * @param obj the object to be referenced
     */
    public PDFReference(PDFObject obj) {
        this.objectNumber = obj.getObjectNumber();
        this.generation = obj.getGeneration();
        this.objReference = new SoftReference<PDFObject>(obj);
    }

    /**
     * Creates a new PDF reference, but without a reference to the original object.
     * @param ref an object reference
     */
    public PDFReference(String ref) {
        if (ref == null) {
            throw new NullPointerException("ref must not be null");
        }
        String[] parts = ref.split(" ");
        assert parts.length == 3;
        this.objectNumber = new PDFObjectNumber(Integer.parseInt(parts[0]));
        this.generation = Integer.parseInt(parts[1]);
        assert "R".equals(parts[2]);
    }

    /**
     * Returns the PDF object
     * @return the PDF object, or null if it has been released
     */
    public PDFObject getObject() {
        if (this.objReference != null) {
            PDFObject obj = this.objReference.get();
            if (obj == null) {
                this.objReference = null;
            }
            return obj;
        } else {
            return null;
        }
    }

    /**
     * Returns the object number.
     * @return the object number
     */
    public PDFObjectNumber getObjectNumber() {
        return this.objectNumber;
    }

    /**
     * Returns the generation.
     * @return the generation
     */
    public int getGeneration() {
        return this.generation;
    }

    /** {@inheritDoc} */
    @Override
    public String toString() {
        StringBuilder textBuffer = new StringBuilder();
        outputInline(null, textBuffer);
        return textBuffer.toString();
    }

    /** {@inheritDoc} */
    public void outputInline(OutputStream out, StringBuilder textBuffer) {
        textBuffer.append(getObjectNumber().getNumber()).append(' ').append(getGeneration()).append(" R");
    }

}

File: dotnet/src/Fop.Pdf/PDFReference.cs
using System.Globalization;
using System.Text;

namespace Fop.Pdf;

/// <summary>
/// A PDF object reference. The reference holds a weak reference to the actual PDF object so the
/// garbage collector can free it if it is not referenced elsewhere; the important part is the
/// reference information (object number + generation) to the object in the PDF file.
/// </summary>
public class PDFReference : IPDFWritable
{
    private readonly PDFObjectNumber _objectNumber;
    private readonly int _generation;

    // Soft reference equivalent: lets the GC reclaim the target object when unreferenced.
    private WeakReference<PDFObject>? _objReference;

    /// <summary>Creates a new PDF reference to the given object.</summary>
    public PDFReference(PDFObject obj)
    {
        _objectNumber = obj.ObjectNumber;
        _generation = obj.Generation;
        _objReference = new WeakReference<PDFObject>(obj);
    }

    /// <summary>
    /// Creates a new PDF reference from an object reference string (e.g. <c>"8 0 R"</c>), without
    /// a reference to the original object.
    /// </summary>
    public PDFReference(string reference)
    {
        ArgumentNullException.ThrowIfNull(reference);
        string[] parts = reference.Split(' ');
        if (parts.Length != 3 || parts[2] != "R")
        {
            throw new ArgumentException("Invalid PDF reference: " + reference, nameof(reference));
        }
        _objectNumber = new PDFObjectNumber(int.Parse(parts[0], CultureInfo.InvariantCulture));
        _generation = int.Parse(parts[1], CultureInfo.InvariantCulture);
    }

    /// <summary>Returns the referenced PDF object, or <see langword="null"/> if it has been released.</summary>
    public PDFObject? Object
    {
        get
        {
            if (_objReference is null)
            {
                return null;
            }
            if (_objReference.TryGetTarget(out PDFObject? obj))
            {
                return obj;
            }
            _objReference = null;
            return null;
        }
    }

    /// <summary>Returns the object number.</summary>
    public PDFObjectNumber ObjectNumber => _objectNumber;

    /// <summary>Returns the generation.</summary>
    public int Generation => _generation;

    /// <inheritdoc/>
    public override string ToString()
    {
        var textBuffer = new StringBuilder();
        OutputInline(null, textBuffer);
        return textBuffer.ToString();
    }

    /// <inheritdoc/>
    public void OutputInline(Stream? output, StringBuilder textBuffer) =>
        textBuffer.Append(ObjectNumber.Number)
            .Append(' ')
            .Append(Generation)
            .Append(" R");
}
